In [2]:
!pip install pymongo pandas a

  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 12.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 313.6/313.6 kB 15.2 MB/s eta 0:00:00
  Created wheel for a: filename=a-1.0-py3-none-any.whl size=1153 sha256=7db52bdf54d2b6ce9c7426473e53eed1d47c5b52a1bbec2544e6e358de9f8b3f
  Stored in directory: /root/.cache/pip/wheels/67/80/51/f3aacec840b7fa425571209754215cbaa104087d5b6e2f7569
Successfully built a


In [3]:
import pandas as pd
from pymongo import MongoClient
uri = "mongodb+srv://group_1:1234567890@cluster0.0kxynkc.mongodb.net/"
client = MongoClient(uri)
db = client["final_project"]
print(db.list_collection_names())

['alz_data']


In [4]:
print("Collections:", db.list_collection_names())

Collections: ['alz_data']


In [5]:
collection = db["alz_data"]
data = list(collection.find())

In [6]:
df = pd.DataFrame(data)
df.head()

_id  PatientID  Age  Gender  Ethnicity  \
0  67ec77f40f613a8cd81cde79       4751   73       0          0   
1  67ec77f40f613a8cd81cde7a       4752   89       0          0   
2  67ec77f40f613a8cd81cde7b       4753   73       0          3   
3  67ec77f40f613a8cd81cde7c       4754   74       1          0   
4  67ec77f40f613a8cd81cde7d       4755   89       0          0   

   EducationLevel        BMI  Smoking  AlcoholConsumption  PhysicalActivity  \
0               2  22.927749        0           13.297218          6.327112   
1               0  26.827681        0            4.542524          7.619885   
2               1  17.795882        0           19.555085          7.844988   
3               1  33.800817        1           12.209266          8.428001   
4               0  20.716974        0           18.454356          6.310461   

   ...  MemoryComplaints  BehavioralProblems       ADL  Confusion  \
0  ...                 0                   0  1.725883          0   
1  ...                 0                   0  2.592424          0   
2  ...                 0                   0  7.119548          0   
3  ...                 0                   1  6.481226          0   
4  ...                 0                   0  0.014691          0   

   Disorientation  PersonalityChanges  DifficultyCompletingTasks  \
0               0                   0                          1   
1               0                   0                          0   
2               1                   0                          1   
3               0                   0                          0   
4               0                   1                          1   

   Forgetfulness  Diagnosis  DoctorInCharge  
0              0          0       XXXConfid  
1              1          0       XXXConfid  
2              0          0       XXXConfid  
3              0          0       XXXConfid  
4              0          0       XXXConfid  

[5 rows x 36 columns]

In [8]:
print(df.columns.tolist())

['PatientID', 'Age', 'Gender', 'Ethnicity', 'EducationLevel', 'BMI', 'Smoking', 'AlcoholConsumption', 'PhysicalActivity', 'DietQuality', 'SleepQuality', 'FamilyHistoryAlzheimers', 'CardiovascularDisease', 'Diabetes', 'Depression', 'HeadInjury', 'Hypertension', 'SystolicBP', 'DiastolicBP', 'CholesterolTotal', 'CholesterolLDL', 'CholesterolHDL', 'CholesterolTriglycerides', 'MMSE', 'FunctionalAssessment', 'MemoryComplaints', 'BehavioralProblems', 'ADL', 'Confusion', 'Disorientation', 'PersonalityChanges', 'DifficultyCompletingTasks', 'Forgetfulness', 'Diagnosis', 'DoctorInCharge']


In [9]:
X = df.drop(['Diagnosis', 'PatientID', 'DoctorInCharge'], axis=1)  # Drop ID and non-relevant columns
y = df['Diagnosis']

In [10]:
# Checking categorical variables
categorical_cols = X.select_dtypes(include=['object']).columns
print("Categorical Columns:", categorical_cols.tolist())

# One-hot encoding categorical variables
X_encoded = pd.get_dummies(X, columns=categorical_cols, drop_first=True)

# Encoding target
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
y_encoded = le.fit_transform(y)

# Splitting the data
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    X_encoded, y_encoded, test_size=0.2, random_state=42
)

# Standardizing numeric features
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

Categorical Columns: []


In [11]:
print(f"X_train shape: {X_train_scaled.shape}")
print(f"X_test shape: {X_test_scaled.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")

X_train shape: (1719, 32)
X_test shape: (430, 32)
y_train shape: (1719,)
y_test shape: (430,)


In [12]:
print(pd.DataFrame(X_train_scaled).head())

         0         1         2         3         4         5         6   \
0  1.328523  0.999418  1.309058 -0.313363  0.010659  1.576964  1.136669   
1 -0.547367 -1.000582 -0.701908  1.898840  1.305827 -0.634130 -1.492106   
2  0.776791  0.999418  2.314541  0.792738 -1.634359 -0.634130  1.099593   
3  0.114712  0.999418 -0.701908  0.792738  0.362836  1.576964  0.317867   
4 -1.540486 -1.000582 -0.701908  0.792738 -0.433186  1.576964 -1.334000   

         7         8         9   ...        22        23        24        25  \
0 -1.648973  0.327959  0.454049  ...  1.234392 -0.686652 -0.509255 -0.430717   
1  1.521599 -1.087222 -1.035466  ... -0.746324  0.189349 -0.509255  2.321709   
2 -1.070683  0.563772 -0.507425  ...  0.729927 -0.486998 -0.509255 -0.430717   
3  0.473723  0.424829 -0.121121  ...  1.606083 -0.145887 -0.509255 -0.430717   
4  1.781140 -1.005189 -1.321369  ... -1.402005  1.568230  1.963653 -0.430717   

         26        27        28        29        30        31  
0  0

In [13]:
import numpy as np
print("Encoded target classes:", np.unique(y_train))

Encoded target classes: [0 1]


In [14]:
!pip install keras-tuner tensorflow

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 129.1/129.1 kB 3.9 MB/s eta 0:00:00


In [15]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.optimizers import Adam
import keras_tuner as kt

In [16]:
def build_model(hp):
    model = Sequential()

    # Tune number of layers (from 1 to 5)
    for i in range(hp.Int('num_layers', 1, 5)):
        model.add(Dense(
            units=hp.Int(f'units_{i}', min_value=16, max_value=256, step=16),
            activation=hp.Choice('activation', values=['relu', 'tanh', 'sigmoid', 'elu', 'selu'])
        ))
        # Optional dropout to prevent overfitting
        model.add(Dropout(hp.Float('dropout', 0.0, 0.5, step=0.1)))

    # Output layer
    model.add(Dense(1, activation='sigmoid'))

    # Compile the model with a tuned learning rate
    model.compile(
        optimizer=Adam(hp.Float('learning_rate', 1e-4, 1e-2, sampling='log')),
        loss='binary_crossentropy',
        metrics=['accuracy']
    )

    return model

In [17]:
tuner = kt.Hyperband(
    build_model,
    objective='val_accuracy',
    max_epochs=50,
    factor=3,
    directory='hyperband_alz',
    project_name='alz_tuning'
)

# Early stopping to avoid unnecessary computation
stop_early = tf.keras.callbacks.EarlyStopping(monitor='val_loss', patience=5)

tuner.search(X_train_scaled, y_train, epochs=50, validation_split=0.2, callbacks=[stop_early])

Trial 90 Complete [00h 00m 09s]
val_accuracy: 0.819767415523529

Best val_accuracy So Far: 0.8459302186965942
Total elapsed time: 00h 08m 30s


In [18]:
# Retrieve the best hyperparameters
best_hps = tuner.get_best_hyperparameters(num_trials=1)[0]

print(f"""
🏅 Optimal Hyperparameters found:
---------------------------------
- Activation Function: {best_hps.get('activation')}
- Number of Layers: {best_hps.get('num_layers')}
- Learning Rate: {best_hps.get('learning_rate'):.5f}
- Dropout Rate: {best_hps.get('dropout'):.2f}
""")

for i in range(best_hps.get('num_layers')):
    print(f"  - Units in Layer {i+1}: {best_hps.get(f'units_{i}')}")


🏅 Optimal Hyperparameters found:
---------------------------------
- Activation Function: relu
- Number of Layers: 5
- Learning Rate: 0.00034
- Dropout Rate: 0.30

  - Units in Layer 1: 80
  - Units in Layer 2: 32
  - Units in Layer 3: 80
  - Units in Layer 4: 80
  - Units in Layer 5: 96


In [19]:
# Build the final model with optimal hyperparameters
final_model = tuner.hypermodel.build(best_hps)

# Train the final model
history = final_model.fit(
    X_train_scaled, y_train,
    epochs=50,
    validation_split=0.2,
    callbacks=[tf.keras.callbacks.EarlyStopping(monitor='val_loss', patience=5)]
)

Epoch 1/50
43/43 ━━━━━━━━━━━━━━━━━━━━ 4s 10ms/step - accuracy: 0.5508 - loss: 0.6856 - val_accuracy: 0.6192 - val_loss: 0.6682
Epoch 2/50
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.6564 - loss: 0.6527 - val_accuracy: 0.6192 - val_loss: 0.6629
Epoch 3/50
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.6517 - loss: 0.6515 - val_accuracy: 0.6192 - val_loss: 0.6563
Epoch 4/50
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.6536 - loss: 0.6528 - val_accuracy: 0.6192 - val_loss: 0.6501
Epoch 5/50
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.6391 - loss: 0.6507 - val_accuracy: 0.6192 - val_loss: 0.6426
Epoch 6/50
43/43 ━━━━━━━━━━━━━━━━━━━━ 1s 7ms/step - accuracy: 0.6711 - loss: 0.6209 - val_accuracy: 0.6192 - val_loss: 0.6291
Epoch 7/50
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.6556 - loss: 0.6185 - val_accuracy: 0.6483 - val_loss: 0.6134
Epoch 8/50
43/43 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.6781 - loss: 0.5979 - val_accuracy: 0.6948 - val_loss

In [20]:
loss, accuracy = final_model.evaluate(X_test_scaled, y_test)
print(f"🎯 Final Test Accuracy: {accuracy:.4f}")

14/14 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.8283 - loss: 0.3905
🎯 Final Test Accuracy: 0.8233


📌 Conclusions:

After completing the hyperparameter optimization study using Keras Tuner with the Hyperband algorithm on your Alzheimer’s disease dataset, we obtained the following key findings:
🥇 Optimal Hyperparameters Identified:
	•	Best Activation Function: ReLU
	•	Optimal Number of Hidden Layers: 5 layers
	•	Optimal Learning Rate: 0.00034
	•	Optimal Dropout Rate: 0.30

Optimal Neurons per Layer:
	•	Layer 1: 80 neurons
	•	Layer 2: 32 neurons
	•	Layer 3: 80 neurons
	•	Layer 4: 80 neurons
	•	Layer 5: 96 neurons

🎯 Model Performance:
	•	Best Validation Accuracy: 84.59%
	•	Final Test Accuracy: 82.33%

The results indicate the optimized model effectively captures relevant patterns in the dataset, achieving solid predictive power in classifying Alzheimer’s disease, with accuracy above 82% on unseen test data.

📊 Practical Observations:
	•	Deep Structure:
The deeper architecture (5 hidden layers) demonstrated effectiveness, suggesting that complex relationships relevant to Alzheimer’s diagnosis are better captured by deeper neural networks.
	•	Dropout:
The optimal dropout rate (30%) effectively reduced overfitting, ensuring the model generalized well to new, unseen data.
	•	Low Learning Rate:
The identified small learning rate (0.00034) allowed stable, incremental updates of model weights, avoiding erratic fluctuations during training and improving overall convergence.
	•	ReLU Activation:
The ReLU activation function performed best, aligning with common best practices in neural networks, efficiently handling non-linear patterns in this classification task.

🎯 Insights from the Alzheimer’s Disease Prediction Study

1. Feasibility of Prediction:
	•	With a test accuracy of ~82.33%, the optimized neural network demonstrates strong potential for accurately predicting Alzheimer’s risk from patient data. This confirms that meaningful patterns exist within your dataset that can be leveraged to make useful predictions.
2. Important Predictors:
	•	Although the current neural network acts as a “black box,” the significant accuracy indicates some of the provided variables—such as age, BMI, cardiovascular health, cholesterol levels, cognitive assessments (like MMSE), and behavioral indicators (e.g., forgetfulness, confusion)—play critical roles in predicting Alzheimer’s.
3. Practical Utility:
	•	Such predictive models, once validated further, could become a valuable clinical tool, potentially allowing for early detection or early-stage risk assessment.
	•	Early prediction may improve patient outcomes by enabling timely medical interventions and lifestyle modifications.
✅ General Conclusion:
The implemented methodology effectively achieved its goal, clearly identifying the neural architecture and hyperparameters that optimally discriminate Alzheimer’s diagnosis within the analyzed dataset.

In [21]:
# Save to .h5 file
final_model.save("alzheimers_model_optimized.h5")